# Intermediate SQL

Here you can access every table used in the course. To access each table, you will need to specify the `cinema` schema in your queries (e.g., `cinema.reviews` for the `reviews` table.

---
_Note: When using sample databases such as those that contain course data, you have read-only access. You can run queries, but cannot make any changes such as adding, deleting, or modifying the data (e.g., creating tables, views, etc.)._

## Take Notes

Add notes about the concepts you've learned and SQL cells with queries you want to keep.

In [1]:
-- Add your own queries here
SELECT *
FROM cinema.reviews
LIMIT 5

id  film_id  num_user  num_critic  imdb_score  num_votes  facebook_likes
0   1     3934       588         432         7.1     203461           46000
1   2     3405       285         267         6.4     149998               0
2   3      478        65          29         3.2       8465             491
3   4       74        83          25         7.6       7071             930
4   5     1254      1437         224         8.0     241030           13000

## Explore Datasets
Use the `descriptions`, `films`, `people`, `reviews`, and `roles` tables to explore the data and practice your skills!
- Which titles in the `reviews` table have an IMDB score higher than 8.5?
- Select all titles from Germany released after 2010 from the `films` table.
- Calculate a count of all movies by country using the `films` table.

# COUNT()

- Example:

```{SQL}
SELECT COUNT(birthdate) AS count_birthdates
FROM people;
```

- With multiple fields:
```{SQL}
SELECT COUNT(name) AS count_names, COUNT(birthdate) AS count_birthdates
FROM people;
```

- Total counts in a table
```{SQL}
SELECT COUNT(*) AS total_records
FROM people;
```


# DISTINCT

In [3]:
SELECT language
FROM cinema.films 
limit 5;

language
0         
1         
2         
3   German
4   German

In [4]:
SELECT DISTINCT language 
FROM cinema.films;

language
0       Danish
1        Greek
2     Dzongkha
3         None
4        Tamil
5      Swahili
6     Mandarin
7         Urdu
8     Filipino
9      Bosnian
10        Maya
11       Dutch
12        Dari
13     Spanish
14    Japanese
15   Slovenian
16      Arabic
17      French
18            
19      Hebrew
20     Aramaic
21   Hungarian
22        Thai
23     Kannada
24      Telugu
25       Czech
26      Kazakh
27      Korean
28   Icelandic
29      Polish
30     Russian
31     Persian
32      German
33     Italian
34     Chinese
35   Mongolian
36   Norwegian
37  Aboriginal
38  Indonesian
39     Swedish
40     Panjabi
41  Vietnamese
42   Cantonese
43  Portuguese
44       Hindi
45    Romanian
46     English
47        Zulu

# COUNT() with DISTINCT

In [6]:
select count(distinct birthdate) as count_distinct_birthdates
from cinema.people;

count_distinct_birthdates
0                       5398

Temos 5398 datas de nascimento distintas.

# Exercícios

> **1. Learning to COUNT()**

In [7]:
SELECT count(film_id) as count_film_id 
FROM cinema.reviews

count_film_id
0           4968

> **2. Practice with COUNT()**

As you've seen, COUNT(*) tells you how many records are in a table. However, if you want to count the number of non-missing values in a particular field, you can call COUNT() on just that field.

1. Count the total number of records in the people table, aliasing the result as count_records.

In [9]:
-- Count the number of records in the people table
SELECT COUNT(*) AS count_records
FROM cinema.people;

count_records
0           8397

2. Count the number of records with a birthdate in the people table, aliasing the result as count_birthdate.

In [10]:
SELECT COUNT(birthdate) AS count_birthdate
FROM cinema.people;

count_birthdate
0             6152

3. Count the records for languages and countries in the `films` table; alias as `count_languages` and `count_countries`.

In [11]:
SELECT *
FROM cinema.films
limit 5;

id                                             title  ...      gross     budget
0   1  Intolerance: Love's Struggle Throughout the Ages  ...        NaN   385907.0
1   2                    Over the Hill to the Poorhouse  ...  3000000.0   100000.0
2   3                                    The Big Parade  ...        NaN   245000.0
3   4                                        Metropolis  ...    26435.0  6000000.0
4   5                                     Pandora's Box  ...     9950.0        NaN

[5 rows x 9 columns]

In [12]:
SELECT COUNT(language) AS count_languages, COUNT(country) AS count_countries
FROM cinema.films;

count_languages  count_countries
0             4968             4968

> **3. SELECT DISTINCT**

Often query results will include many duplicate values. You can use the DISTINCT keyword to select the unique values from a field.

This might be useful if, for example, you're interested in knowing which languages are represented in the films table. See if you can find out what countries are represented in this table with the following exercises.

1. Return the unique countries represented in the films table using DISTINCT.

In [14]:
SELECT DISTINCT(country)
FROM cinema.films;

country
0     Soviet Union
1        Indonesia
2            Italy
3         Cameroon
4   Czech Republic
..             ...
60         Belgium
61          Mexico
62          Poland
63          Taiwan
64        Thailand

[65 rows x 1 columns]

2. Return the number of unique countries represented in the films table, aliased as count_distinct_countries.

In [16]:
SELECT COUNT(DISTINCT(country)) AS count_distinct_countries 
FROM cinema.films;

count_distinct_countries
0                        65

# SQL style

- Não precisa identar, mas é uma boa prática para tornar a query mais legível. 
- https://www.sqlstyle.guide/

In [17]:
SELECT 
	title,
	release_year,
	country 
FROM cinema.films
LIMIT 3;

title  release_year country
0  Intolerance: Love's Struggle Throughout the Ages          1916     USA
1                    Over the Hill to the Poorhouse          1920     USA
2                                    The Big Parade          1925     USA

# WHERE

In [1]:
SELECT * 
FROM cinema.films
LIMIT 5;

id                                             title  ...      gross     budget
0   1  Intolerance: Love's Struggle Throughout the Ages  ...        NaN   385907.0
1   2                    Over the Hill to the Poorhouse  ...  3000000.0   100000.0
2   3                                    The Big Parade  ...        NaN   245000.0
3   4                                        Metropolis  ...    26435.0  6000000.0
4   5                                     Pandora's Box  ...     9950.0        NaN

[5 rows x 9 columns]

In [2]:
SELECT title
FROM cinema.films
WHERE release_year > 1960;

title
0                           Judgment at Nuremberg
1                           Pocketful of Miracles
2                                     The Hustler
3                                     The Misfits
4                                 West Side Story
...                                           ...
4844                                     Warcraft
4845                            X-Men: Apocalypse
4846  Xi you ji zhi: Sun Wukong san da Baigu Jing
4847                                  Yoga Hosers
4848                                  Zoolander 2

[4849 rows x 1 columns]

In [3]:
SELECT title
FROM cinema.films
WHERE release_year <= 1960;

title
0   Intolerance: Love's Struggle Throughout the Ages
1                     Over the Hill to the Poorhouse
2                                     The Big Parade
3                                         Metropolis
4                                      Pandora's Box
..                                               ...
72                                  Some Like It Hot
73                                   The Nun's Story
74                                      Elmer Gantry
75                                            Psycho
76                                     The Apartment

[77 rows x 1 columns]

In [4]:
-- not equal to
SELECT title
FROM cinema.films
WHERE release_year <> 1960;

title
0     Intolerance: Love's Struggle Throughout the Ages
1                       Over the Hill to the Poorhouse
2                                       The Big Parade
3                                           Metropolis
4                                        Pandora's Box
...                                                ...
4918                                          Warcraft
4919                                 X-Men: Apocalypse
4920       Xi you ji zhi: Sun Wukong san da Baigu Jing
4921                                       Yoga Hosers
4922                                       Zoolander 2

[4923 rows x 1 columns]

In [5]:
-- WHERE with strings
SELECT title
FROM cinema.films
WHERE country = 'Japan';

title
0             Seven Samurai
1         Tora! Tora! Tora!
2                     Akira
3                  Madadayo
4            Street Fighter
5    The Quick and the Dead
6         Princess Mononoke
7                  The Cure
8             Godzilla 2000
9      Pokémon 3: The Movie
10            Spirited Away
11     Howl's Moving Castle
12                 Steamboy
13          One Missed Call
14                    Ponyo
15    Harlock: Space Pirate
16                     R100
17               Trainwreck
18      Godzilla Resurgence
19      Godzilla Resurgence
20  The Secret Life of Pets
21  Yu-Gi-Oh! Duel Monsters

# Exercícios

> **1. Using WHERE with numbers**

Filtering with WHERE allows you to analyze your data better. You may have a dataset that includes a range of different movies, and you need to do a case study on the most notable films with the biggest budgets. In this case, you'll want to filter your data to a specific budget range.

Now it's your turn to use the WHERE clause to filter numeric values!

**1. Select the film_id and imdb_score from the reviews table and filter on scores higher than 7.0.**

In [6]:
SELECT * 
FROM cinema.reviews
LIMIT 5;

id  film_id  num_user  num_critic  imdb_score  num_votes  facebook_likes
0   1     3934       588         432         7.1     203461           46000
1   2     3405       285         267         6.4     149998               0
2   3      478        65          29         3.2       8465             491
3   4       74        83          25         7.6       7071             930
4   5     1254      1437         224         8.0     241030           13000

In [7]:
-- Select film_ids and imdb_score with an imdb_score over 7.0
SELECT film_id, imdb_score
FROM cinema.reviews
WHERE imdb_score > 7.0

film_id  imdb_score
0        3934         7.1
1          74         7.6
2        1254         8.0
3        4841         8.1
4        3252         7.2
...       ...         ...
1531      199         8.0
1532     1814         7.2
1533     4158         8.0
1534     4086         7.1
1535       29         8.6

[1536 rows x 2 columns]

**2. Select the film_id and facebook_likes of the first ten records with less than 1000 likes from the reviews table.**

In [9]:
-- Select film_ids and facebook_likes for ten records with less than 1000 likes 
SELECT film_id, facebook_likes
FROM cinema.reviews
WHERE facebook_likes < 1000
LIMIT 10;

film_id  facebook_likes
0     3405               0
1      478             491
2       74             930
3      740               0
4     2869             689
5     1181               0
6     2020               0
7     2312             912
8     1820             872
9      831             975

**3. Count how many records have a num_votes of at least 100,000; use the alias films_over_100K_votes.**

In [10]:
SELECT COUNT(*) AS films_over_100K_votes
FROM cinema.reviews
WHERE num_votes >= 100000;

films_over_100k_votes
0                   1211

> **2. Using WHERE with text**
> 
WHERE can also filter string values.

Imagine you are part of an organization that gives cinematography awards, and you have several international categories. Before you confirm an award for every language listed in your dataset, it may be worth seeing if there are enough films of a specific language to make it a fair competition. If there is only one movie or a significant skew, it may be worth considering a different way of giving international awards.

- Select and count the language field using the alias count_spanish.
- Apply a filter to select only Spanish from the language field.

In [1]:
-- Count the Spanish-language films
SELECT COUNT(language) as count_spanish 
FROM cinema.films
WHERE language = 'Spanish'


count_spanish
0             40

# Multiple criteria

- OR
- AND
- BETWEEN

In [2]:
SELECT title 
FROM cinema.films
WHERE cinema.films.release_year = 1994
	OR cinema.films.release_year = 2000;


title
0            3 Ninjas Kick Back
1        A Low Down Dirty Shame
2    Ace Ventura: Pet Detective
3                Baby's Day Out
4         Beverly Hills Cop III
..                          ...
220                     Whipped
221                Woman on Top
222                 Wonder Boys
223                       X-Men
224         You Can Count on Me

[225 rows x 1 columns]

In [3]:
SELECT title 
FROM cinema.films
WHERE cinema.films.release_year > 1994
	OR cinema.films.release_year < 2000;

title
0     Intolerance: Love's Struggle Throughout the Ages
1                       Over the Hill to the Poorhouse
2                                       The Big Parade
3                                           Metropolis
4                                        Pandora's Box
...                                                ...
4921                                          Warcraft
4922                                 X-Men: Apocalypse
4923       Xi you ji zhi: Sun Wukong san da Baigu Jing
4924                                       Yoga Hosers
4925                                       Zoolander 2

[4926 rows x 1 columns]

In [4]:
SELECT title 
FROM cinema.films
WHERE (release_year = 1994 OR release_year = 1994)
	AND (certification = 'PG'OR certification = 'R');

title
0                                  3 Ninjas Kick Back
1                              A Low Down Dirty Shame
2                                      Baby's Day Out
3                               Beverly Hills Cop III
4                               Bullets Over Broadway
5                                              Clerks
6                                          Disclosure
7                                             Ed Wood
8                                             Exotica
9                         Four Weddings and a Funeral
10                                 Heavenly Creatures
11  Interview with the Vampire: The Vampire Chroni...
12                                Legends of the Fall
13                                       Little Women
14                                         Men of War
15                                      New Nightmare
16                                   On Deadly Ground
17                  Police Academy: Mission to Moscow
18                                       Pulp Fiction
19                                           Rapa Nui
20                                        Ri¢hie Ri¢h
21                                         Serial Mom
22                                              Speed
23                             Star Trek: Generations
24                                           The Crow
25                                    The Flintstones
26                                The Hudsucker Proxy
27                                        The Inkwell
28                       The Legend of Drunken Master
29                                   The Santa Clause
30                           The Shawshank Redemption
31                                     The Specialist
32                                 There Goes My Baby
33                                            Timecop
34                                          True Lies
35                                               Wolf

In [5]:
SELECT title
FROM cinema.films
WHERE cinema.films.release_year		
	BETWEEN 1994 AND 2000;

title
0            3 Ninjas Kick Back
1        A Low Down Dirty Shame
2    Ace Ventura: Pet Detective
3                Baby's Day Out
4         Beverly Hills Cop III
..                          ...
809                     Whipped
810                Woman on Top
811                 Wonder Boys
812                       X-Men
813         You Can Count on Me

[814 rows x 1 columns]

In [6]:
SELECT title
FROM cinema.films
WHERE cinema.films.release_year
BETWEEN 1994 AND 2000 AND country = 'UK';

title
0     Four Weddings and a Funeral
1             The Hudsucker Proxy
2                Dead Man Walking
3                       GoldenEye
4                     Richard III
..                            ...
59                         Snatch
60                      The Claim
61                      The Claim
62             The House of Mirth
63  Thomas and the Magic Railroad

[64 rows x 1 columns]

# Exercícios

> **Using AND**

The following exercises combine AND and OR with the WHERE clause. Using these operators together strengthens your queries and analyses of data.

You will apply these new skills now on the films database.

In [7]:
-- Select the title and release_year for all German-language films released before 2000
SELECT title,
       release_year
FROM cinema.films
WHERE language = 'German' AND release_year < 2000;

title  release_year
0                         Metropolis          1927
1                      Pandora's Box          1929
2  The Torture Chamber of Dr. Sadism          1967
3                           Das Boot          1981
4                       Run Lola Run          1998
5                     Aimee & Jaguar          1999

In [8]:
-- Select all records for German-language films released after 2000 and before 2010
SELECT *
FROM cinema.films
WHERE language = 'German' 
    AND (release_year > 2000 AND release_year < 2010);

id                       title  ...       gross    budget
0  1952             Good Bye Lenin!  ...   4063859.0   4800000
1  2130                    Downfall  ...   5501940.0  13500000
2  2224                Summer Storm  ...     95016.0   2700000
3  2709         The Lives of Others  ...  11284657.0   2000000
4  3100  The Baader Meinhof Complex  ...    476270.0  20000000
5  3143                    The Wave  ...         NaN   5000000
6  3220                       Cargo  ...         NaN   4500000
7  3346                Soul Kitchen  ...    274385.0   4000000
8  3412            The White Ribbon  ...   2222647.0  12000000

[9 rows x 9 columns]

> **Using OR**

This time you'll write a query to get the title and release_year of films released in 1990 or 1999, which were in English or Spanish and took in more than $2,000,000 gross.

It looks like a lot, but you can build the query up one step at a time to get comfortable with the underlying concept in each step. Let's go!


1. Select the title and release_year for films released in 1990 or 1999 using only WHERE and OR.
2. Filter the records to only include English or Spanish-language films.
3. Finally, restrict the query to only return films worth more than $2,000,000 gross.

In [11]:
-- Find the title and year of films from the 1990 or 1999
SELECT 
	title,
	release_year
FROM cinema.films
WHERE release_year = 1990 OR release_year = 1999 
	AND (language = 'English' OR language = 'Spanish')
	AND gross > 2000000;
	

title  release_year
0                    Arachnophobia          1990
1      Back to the Future Part III          1990
2                   Child's Play 2          1990
3               Dances with Wolves          1990
4                  Days of Thunder          1990
..                             ...           ...
167                       Trippin'          1999
168  Universal Soldier: The Return          1999
169                  Varsity Blues          1999
170                 Wild Wild West          1999
171                 Wing Commander          1999

[172 rows x 2 columns]

> **Using BETWEEN**

Let's use BETWEEN with AND on the films database to get the title and release_year of all Spanish-language (or French) films released between 1990 and 2000 (inclusive) with budgets over $100 million.


In [12]:
SELECT title, release_year
FROM cinema.films
WHERE release_year BETWEEN 1990 AND 2000
	AND budget > 100000000
	AND (language = 'Spanish' OR language = 'French');

title  release_year
0  Les couloirs du temps: Les visiteurs II          1998
1                                    Tango          1998